Heap profiling

fdkong opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

Closes #15990

fdkong committed:
Changed MoosApp.{C,h} to take head profilling

Closes #15990

fdkong committed:
Added docs for heap profiling

lindsayad commented on modules/doc/content/application_development/profiling.md:
````diff
@@ -47,6 +47,22 @@ This will use the filename base you pass and append a suffix of the form
+Similarly, you can do a heap profiling like this:
+
+```
+HEAP_PROFILE_INUSE_INTERVAL=104857600 MOOSE_HEAP_BASE=run1_ mpiexec -n 32 ./your-app_oprof -i input_file.i
````
Isn't this the default value? Does anything happen if you don't specify this environment value? I remember when I was trying to do this, I would not get any memory profiles dumped ever, but I wasn't specifying any environment variable values like `HEAP_PROFILE_INUSE_INTERVAL`

fdkong replied:
The default is nothing. Right, you do not get anything if you do not specify `HEAP_PROFILE_INUSE_INTERVAL` or other variables

fdkong committed:
Fixed bugs

fdkong commented:
Do not merge it now. I am going to try it out

lindsayad commented on framework/src/base/MooseApp.C:
```diff
@@ -343,13 +344,29 @@ MooseApp::MooseApp(InputParameters parameters)
+  if (std::getenv("MOOSE_HEAP_BASE"))
+  {
+    static std::string profile_file =
+        std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank()) + ".prof";
```
I don't think we want the `.prof` string in here do we? At least on my machine, libtcmalloc actually appends a `.heap` suffix on its own

fdkong replied:
Ok, I will remove that

fdkong replied:
Done

lindsayad committed:
Ensure that profiling libs come before libs from libmesh

I had a heck of a time figuring out why I wasn't getting heap profiles,
and the reason is that `libc` came before `libtcmalloc` on my link line,
e.g. malloc calls were going straight to `libc` instead of through
`libtcmalloc`. So first of all we should always be putting `LDFLAGS`
before `LIBS`. And we are definitely slightly abusing the meaning of
`LDFLAGS` by putting the profiler libs in `LDFLAGS` but I'm not too bothered by
that. Also make sure that we insert the profiler libs in the front of
`libmesh_LDFLAGS` in a further attempte to make sure we're giving those
profiler library symbols as high a priority as possible

lindsayad commented:
This is working really well for me after the `.mk` fixes. I love this. Thanks @fdkong

fdkong committed:
Drop ".prof" for heap profiling because ".heap" will be applied automatically

fdkong commented:
OK, it works for me on HPC.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/626696/) on bce6442 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16224/site)

This comment will be updated on new commits.

lindsayad reviewed (commented):
I have a commit on this now, so maybe we should have another reviewer

fdkong commented:
@rwcarlsen Could you take a look? Thanks

lindsayad commented:
Something I found today is that if tcmalloc symbols are ahead of libc symbols, then you get a bazillion valgrind errors when running valgrind (invalid frees, etc.). So that is unfortunate

YaqiWang commented on modules/doc/content/application_development/profiling.md:
````diff
@@ -47,6 +47,22 @@ This will use the filename base you pass and append a suffix of the form
+
+```
+HEAP_PROFILE_INUSE_INTERVAL=104857600 MOOSE_HEAP_BASE=run1_ mpiexec -n 32 ./your-app_oprof -i input_file.i
+```
````
Maybe we are missing an empty line bellow? Check this out; https://mooseframework.inl.gov/docs/PRs/16224/site/application_development/profiling.html.

fdkong replied:
Thanks. Good catch. I will fix that
